openhtf/__init__.py

```diff
@@ -84,32 +84,56 @@ def __call__(self, test_record):  # pylint: disable=invalid-name
       f.write(self.encode(as_dict))
 
 
-def TestPhase(timeout_s=None, run_if=None):  # pylint: disable=invalid-name
-  """Decorator to wrap a test phase function with the given options.
-
-  Args:
+class TestPhaseOptions(mutablerecords.Record(
+    'TestPhaseOptions', [], {'timeout_s': None, 'run_if': None})):
+  """
+  Attributes:
     timeout_s: Timeout to use for the phase, in seconds.
     run_if: Callback that decides whether to run the phase or not.  The
       callback will be passed the phase_data the phase would be run with.
 
-  Returns:
-    A wrapper function that takes a phase_func and returns a
-        TestPhaseInfo for it with the given options set.
+  Example Usage:
+    @TestPhaseOptions(timeout_s=1)
+    def PhaseFunc(test):
+      pass
   """
 
-  def Wrap(phase_func):  # pylint: disable=invalid-name
-    """Attach the given options to the phase_func."""
+  def __call__(self, phase_func):
+    phase = TestPhaseInfo.WrapOrReturn(phase_func)
+    phase.options = self
+    return phase
 
-    # Test Phases must take at least one argument (the phase data tuple).
-    if len(inspect.getargspec(phase_func).args) < 1:
-      raise InvalidTestPhaseError(phase_func, 'Not enough args')
+PhasePlug = mutablerecords.Record('PhasePlug', ['name', 'cls', 'update_kwargs'])
 
-    if timeout_s is not None:
-      phase_func.timeout_s = timeout_s
-    if run_if is not None:
-      phase_func.run_if = run_if
-    return phase_func
-  return Wrap
+
+class TestPhaseInfo(mutablerecords.Record(
```

Comment:
Is there a mailing list for this sort of thing or do should we go via bugs? There's nothing about governance or discussion in the CONTRIBUTING.md file. In lieu of that, I talked to Joe in person regarding this.

I could see that it was an explicit change. However, I was doing other things and this started to get in my way so I changed it one piece at a time until I was done and realized I had fully converted it back. I talked to Joe in person after sending this out and he seemed potentially in favor of it, but hadn't seen it just yet.

The thing that bothered me the most was the '.wraps' loops in various places. I also think this cleaned things up and didn't change the user API at all. They're free to add decorators and such to their phases. Before, if they wanted things like plugs and measurements to bubble up, they had to add outer_func.wraps = wrapped_func, which was not documented anywhere. Documenting it as "openhtf.TestPhaseInfo.WrapOrReturn(wrapped_func).func = outer_func" versus the .wraps earlier is also less prone to typo debugging.